src/leo_nlm_proto4_server.erl

```diff
@@ -0,0 +1,314 @@
+%% @doc
+nlmproc4_null_4(_Clnt, State) ->
+    ?debug("NLM_NULL", "from:~p", [_Clnt]),
+    {reply, [], State}.
+
+%% @doc
+nlmproc4_test_4({Cookie, IsExclusive, Lock} = TestArgs, Clnt, State) ->
+    ?debug("NLM_TEST", "Args:~p, from:~p", [TestArgs, Clnt]),
+    {_CallerName, FileHandler, _Owner, _Svid, _Offset, _Length} = Lock,
+    NewLock = convert_lock(Lock, IsExclusive), 
+    LockRet = case test_lock(FileHandler, NewLock) of
+                  {ok, _} ->
+                      ?NLM4_GRANTED;
+                  _ ->
+                      ?NLM4_DENIED
+              end,
+    {reply, {Cookie, {LockRet}}, State}.
+
+nlmproc4_lock_4({Cookie, _IsBlock, IsExclusive, Lock, _IsReclaim, _NSMState} = LockArgs, Clnt, State) ->
+    ?debug("NLM_LOCK", "Args:~p, from:~p", [LockArgs, Clnt]),
+    {_CallerName, FileHandler, _Owner, _Svid, _Offset, _Length} = Lock,
+    NewLock = convert_lock(Lock, IsExclusive), 
+    LockRet = case test_lock(FileHandler, NewLock) of
+                  {ok, ExistingLocks} ->
+                      ets:insert(?NLM_LOCK_ETS, {FileHandler, [NewLock | ExistingLocks]}),
+                      ?NLM4_GRANTED;
+                  _ ->
+                      ?NLM4_DENIED
+              end,
+    {reply, {Cookie, {LockRet}}, State}.
+
+nlmproc4_cancel_4({Cookie, _IsBlock, _IsExclusive, _Lock} = CancArgs, Clnt, State) ->
+    ?debug("NLM_CANCEL", "Args:~p, from:~p", [CancArgs, Clnt]),
+    {reply, {Cookie, {?NLM4_GRANTED}}, State}.
+
+nlmproc4_unlock_4({Cookie, Lock} = UnlockArgs, Clnt, State) ->
+    ?debug("NLM_UNLOCK", "Args:~p, from:~p", [UnlockArgs, Clnt]),
+    {_CallerName, FileHandler, Owner, _Svid, Offset, Length} = Lock,
+    End = Offset + Length - 1,
+    case ets:lookup(?NLM_LOCK_ETS, FileHandler) of
+        [{FileHandler, ExistingLocks}] ->
+            NewLocks = lists:foldl(fun(CurLock, Acc) ->
+                                           case CurLock#lock_record.owner of
+                                               Owner ->
+                                                   case modify_lock(CurLock, Offset, End) of
+                                                       [] ->
+                                                           Acc;
+                                                       Modified ->
+                                                           ?debug("nlmproc4_unlock_4/3", "Modified: ~p", [Modified]),
+                                                           lists:append(Acc, Modified)
+                                                   end;
+                                               _ ->
+                                                   [CurLock | Acc]
+                                           end
+                                   end, [], ExistingLocks),
+            ets:insert(?NLM_LOCK_ETS, {FileHandler, NewLocks});
+        _ ->
+            ok
+    end,
+    {reply, {Cookie, {?NLM4_GRANTED}}, State}.
+
```

Comment:
We need to catch exceptions generated by ets and if caught, 
We should return NLM4_DENIED_NOLOCKS instead of NLM4_DENIED to comply with NLM v4 spec.